File: src/neetcode/Java/_53MaximumSubarray.java
package neetcode.Java;

public class _53MaximumSubarray {
    public int maxSubArray(int[] nums) {
        int maxSum = nums[0];
        int currSum = maxSum;
        for (int i = 1; i < nums.length; i++) {
            if(currSum<0) currSum=0;
            currSum += nums[i];
            if (currSum > maxSum) {
                maxSum = currSum;
            }
        }
        return maxSum;
    }

    public static void main(String[] args) {
        _53MaximumSubarray obj = new _53MaximumSubarray();
        {
            int[] nums = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
            System.out.println(obj.maxSubArray(nums));
        }
        {
            int[] nums = {5, 4, -1, 7, 8};
            System.out.println(obj.maxSubArray(nums));
        }
        {
            int[] nums = {-2, -1};
            System.out.println(obj.maxSubArray(nums));
        }
        {
            int[] nums = {-1, -2};
            System.out.println(obj.maxSubArray(nums));
        }
    }
}

File: src/neetcode/kotlin/_53MaximumSubarray.kt
package neetcode.kotlin

class _53MaximumSubarray {
    fun maxSubArray(nums: IntArray): Int {
        var maxSum = nums[0]
        var currSum = maxSum
        for (i in 1 until nums.size) {
            if (currSum < 0) currSum = 0
            currSum += nums[i]
            if (currSum > maxSum) {
                maxSum = currSum
            }
        }
        return maxSum
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _53MaximumSubarray()
            run {
                val nums = intArrayOf(-2, 1, -3, 4, -1, 2, 1, -5, 4)
                println(obj.maxSubArray(nums))
            }
            run {
                val nums = intArrayOf(5, 4, -1, 7, 8)
                println(obj.maxSubArray(nums))
            }
            run {
                val nums = intArrayOf(-2, -1)
                println(obj.maxSubArray(nums))
            }
            run {
                val nums = intArrayOf(-1, -2)
                println(obj.maxSubArray(nums))
            }
        }
    }
}
